include/toypp/sharedptr.hpp

```diff
@@ -0,0 +1,224 @@
+#ifndef TOYPP_SHAREDPTR_HPP_
+#define TOYPP_SHAREDPTR_HPP_
+
+#include <atomic>
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+namespace tpp {
+
+namespace details {
+template <typename T>
+struct is_unbounded_array : std::false_type {};
+
+template <typename T>
+struct is_unbounded_array<T[]> : std::true_type {};
+}  // namespace details
+
+template <typename T>
+class WeakPtr;
+
+// TODO support static pointer cast conversions
+template <typename T>
+class SharedPtr {
+  using value_type = std::remove_cv_t<std::remove_reference_t<T>>;
+
+  static constexpr auto k_null_deleter = [](value_type*) { /* do nothing. */ };
+  static constexpr auto k_default_deleter = [](value_type* ptr) { delete ptr; };
```

Comment:
The default deleter will be used for bounded array types, which results in incorrectly using `delete` instead of `delete[]`.  (I haven't worked through all the details here, so I might be wrong here.)